apps/site/lib/site_web/views/bus_stop_change_view.ex

```diff
@@ -1,80 +1,42 @@
+  alias Alerts.{Alert, HistoricalAlert}
   alias Stops.Stop
 
-  @type alerts_by_stop :: [{Stop.t(), [%Alert{}]}]
+  defdelegate affected_routes(alert), to: AmbiguousAlert
+  defdelegate related_stops(alert), to: AmbiguousAlert
+  defdelegate alert_start_date(alert), to: AmbiguousAlert
+  defdelegate alert_municipality(alert), to: AmbiguousAlert
 
-  @spec sorted_by_start_date([%Alert{}]) :: [%Alert{}]
-  def sorted_by_start_date(alerts) do
-    Enum.sort_by(
-      alerts,
-      fn %Alert{active_period: [{start_date, _} | _]} ->
-        start_date
-      end
-    )
-  end
-
-  @spec grouped_by_municipality([%Alert{}]) :: %{String.t() => alerts_by_stop}
-  def grouped_by_municipality(alerts) do
-    grouped_by_stop(alerts)
-    |> Enum.group_by(fn
-      {%Stop{municipality: municipality}, _alerts} -> municipality
-      {_stopname, _alerts} -> nil
-    end)
-  end
-
-  @spec grouped_by_stop([%Alert{}]) :: alerts_by_stop
   def grouped_by_stop(alerts) do
     alerts
-    |> Enum.flat_map(fn alert ->
-      alert
-      |> related_stops()
-      |> Enum.map(&{&1, alert})
-    end)
+    |> Enum.flat_map(&alert_with_related_stops/1)
     |> Enum.group_by(&elem(&1, 0), &elem(&1, 1))
-    |> Enum.sort_by(fn
-      {%Stop{name: name}, _alerts} -> name
-      {stopname, _alerts} -> stopname
-    end)
-    |> Enum.map(fn {stop, alerts} ->
-      {stop, sorted_by_start_date(alerts)}
+    |> sort_and_group_with_fn(fn alerts_list ->
+      Enum.sort_by(alerts_list, &alert_start_date/1)
     end)
   end
 
-  @spec affected_routes(%Alert{}) :: [Phoenix.HTML.Safe.t()]
-  def affected_routes(%Alert{} = alert) do
-    Alert.get_entity(alert, :route)
-    |> MapSet.delete(nil)
-    |> MapSet.to_list()
-    |> Enum.uniq()
-    |> Enum.map(fn routeIdOrName ->
-      name =
-        case Routes.Repo.get(routeIdOrName) do
-          %Route{name: name} -> name
-          _ -> routeIdOrName
-        end
-
-      content_tag(:span, name,
-        class: "c-icon__bus-pill--small u-bg--bus",
-        style: "margin-right: .25rem;"
-      )
-    end)
+  defp alert_with_related_stops(alert) do
+    alert
+    |> related_stops()
+    |> Enum.map(&{&1, alert})
+  end
+
+  def grouped_by_municipality(alerts) do
+    alerts
+    |> Enum.group_by(&alert_municipality/1)
+    |> sort_and_group_with_fn(&grouped_by_stop/1)
   end
 
-  @spec related_stops(%Alert{}) :: [%Stop{}]
-  def related_stops(%Alert{} = alert) do
-    Alert.get_entity(alert, :stop)
-    |> MapSet.delete(nil)
-    |> MapSet.to_list()
-    |> Enum.map(fn stopIdOrName ->
-      case Stops.Repo.get_parent(stopIdOrName) do
-        %Stop{} = stop -> stop
-        _ -> stopIdOrName
-      end
+  # Remaps the grouped alerts by a given function, and sorts by group keys.
+  defp sort_and_group_with_fn(grouped_alerts_list, func) do
+    grouped_alerts_list
+    |> Enum.map(fn {key, alerts} ->
+      {key, func.(alerts)}
     end)
-    |> Enum.uniq()
+    |> Enum.sort_by(&elem(&1, 0))
+    |> Enum.into(%{})
```

Comment:
question: can you say more about what you're trying to do here? Maps aren't inherently ordered (and in fact the order is changing in Erlang 26), so if there's an order you're depending on this should be a List.

Reply:
👍🏼   Yeah that doesn't work. Am trying to ensure everything is grouped _and_ sorted, as the template ultimately shows the alerts grouped by municipality, and grouped by stop name within each municipality. I've refactored this and everything's in Lists now.